### DISCUSSION TITLE
request returning a Promise?

### DISCUSSION THREAD

**Discussion by danikaze** [vercel/next.js#45125] [Help]
I've been using this code in my previous projects from long time ago (last one was using next 11)
```ts
export const IsomorphicLogger: IsomorphicLoggerConstructor = IS_SERVER
  ? require('./private/server').ServerLogger
  : require('./private/client').ClientLogger;
```
It was working pretty well, but now I tried installing next13 and it's failing because `require` returns a `Promise`.

Isn't `require` coming from webpack? And wasn't it supposed to be a sync operation?
In this other projects it was only transformed to async if it detected the syntax `require(...).then`

What's changed now?

**Comment by icyJoseph** [comment]
I've tried to investigate this before, and even though I had a reproduction demo, I still found it hard to construct an answer. Pretty much, the problem is that WebPack uses require async when there's ES modules on the other side, or something like that.... https://webpack.js.org/blog/2020-10-10-webpack-5-release/#async-modules - and that was introduced somewhere in WebPack 5...

**Reply by danikaze** [reply]
@icyJoseph weird, since I had `config.webpack5 = true;` enabled in the old next11 project and this was working then...
it basically makes it impossible to have code like that :/

**Reply by icyJoseph** [reply]
You know I just tested this and it worked, but the private folder was at the same level as pages....

**Reply by icyJoseph** [reply]
I remember that to trigger the promise behavior, somehow webpack has to be unable to find the dependency at compile time or something like that... I've seen this issue before, but can't remember/find the discussion.

**Reply by danikaze** [reply]
Thanks for your comment @icyJoseph I still have to try, hopefully this weekend I'll have time... That made me think it might be just that the constant was not so much of a constant but some runtime value in this new setup... need to double check!

**Reply by danikaze** [reply]
Ok I just double checked, it keeps working :) 
The `IS_SERVER` needs to really be a constant at build time, and that's easy to be done by providing a custom webpack config using `next.config.js` like this:
```js
webpack: (config, { isServer, webpack }) => {
  config.plugins = config.plugins || [];
  config.plugins.push(
    new webpack.DefinePlugin({
      IS_SERVER: isServer,
    })
  );
  return config;
},
```

Then things like this work gladly as expected:
```ts
const Router = IS_SERVER
  ? require('react-router-dom/server').StaticRouter
  : require('react-router-dom').BrowserRouter;
```

**Reply by icyJoseph** [reply]
Why not use `nextRuntime` instead? The `isServer` flag is also true in Middleware/Edge. The value of `nextRuntime` is undefined for browser builds.

Also why not just check `typeof window === "undefined"` and use that from a helper file, or just in-lined.

Good that it works though!

**Reply by danikaze** [reply]
You can't use `typeof window` because to achieve that kind of code you cannot rely on runtime values (which probably was the reason why `require` was returning a `Promise`), they need to be constant at build time or webpack won't know how to remove the unneeded code branches.

As you know next executes the build twice, once for the server and once for the client. With the proper constant the code will be _optimized_ and you will have
```ts
// in your source code
const Router = IS_SERVER
  ? require('react-router-dom/server').StaticRouter
  : require('react-router-dom').BrowserRouter;

// in the server build it results on
const Router = require('react-router-dom/server').StaticRouter;

// in the client build it results on 
const Router = require('react-router-dom').BrowserRouter;
```

By the way, if you don't like "magic constants" like the one defined by using `webpack.DefinePlugin`, `process.env.IS_SERVER` can also be used just by having your `.env` file defining it
```ts
// .env
IS_SERVER=1
```
since it's not prefixed with `NEXT_PUBLIC_` it will be `undefined` in client side and will work the same :)

**Reply by icyJoseph** [reply]
Sorry for being a bit of a last word dummy here, but that doesn't make sense. 

```tsx
import Head from "next/head";

const logger =
  typeof window === "undefined"
    ? require("../private/server").ServerLogger
    : require("../private/client").ClientLogger;

export const getStaticProps = () => {
  logger();
  return {
    props: {},
    revalidate: 1,
  };
};

export default function Home() {
  logger();
  return (
    <div>
      <Head>
        <title>Home title</title>
      </Head>

      <h1>home</h1>
    </div>
  );
}
```

With both private server and client looking like:

```ts
export const ClientLogger = () => console.log("Client");
```

and 

```ts
export const ServerLogger = () => console.log("Server");
```

Now I build it, and peek into the built files:

![Screenshot 2023-01-28 at 08 23 51](https://user-images.githubusercontent.com/21013447/215253012-0d10744d-55f1-4ce8-92eb-39e21912300b.png)

The top editor shows that, on the server we have code eliminated the right arm of the ternary, and as you can see, no surprise `typeof window === "undefined"` was replaced by `true`. Which makes sense since this run in Node.js.

The bottom editor shows the client chunk. Notice how, even though, this run on Node.js, it still picked-up the client logger, and also inlined it. Does that mean that at build time, `typeof window === "undefined"` yielded `false`?  Kind of---, but let's add some logs:

```ts
console.log("\nwindow:", typeof window, "\n");
```

And we build again:

![Screenshot 2023-01-28 at 08 32 57](https://user-images.githubusercontent.com/21013447/215253407-4d55ab74-b34d-4375-85fa-09f87cfd127b.png)

This is part of the client bundle making process, where the code is not really run, but it is compiled, minified, etc... and under that, `typeof window` returns "object"!

What _I think_ happened in your case is that the module you are requiring is just more complex, or perhaps both server and client bundles have a common dependency? Or rather, a more complex dependency structure that the regular bundling cannot do?

**Reply by danikaze** [reply]
Yeah... So I was thinking exactly the same.
I don't know why in the first place the require returned a Promise...
Funny thing, is working now. With `typeof window` so not even defining constants nor envvars are required (which was my original understanding)... I wonder what was different when I encountered it... :-/
